cli/internal/cmd/apply_test.go

```diff
@@ -167,5 +177,215 @@ func TestSkipPhases(t *testing.T) {
+			createTfState:      defaultTfState,
+			flags:              applyFlags{},
+			wantPhases: skipPhases{
+				skipInitPhase:  struct{}{},
+				skipImagePhase: struct{}{}, // No image upgrades on QEMU
+			},
+		},
+		"no config file": {
+			createConfig:       func(require *require.Assertions, fh file.Handler) {},
+			createState:        defaultState,
+			createMasterSecret: defaultMasterSecret,
+			createAdminConfig:  defaultAdminConfig,
+			createTfState:      defaultTfState,
+			flags:              applyFlags{},
+			wantErr:            true,
+		},
+		"no admin config file, but mastersecret file exists": {
+			createConfig:       defaultConfig(cloudprovider.GCP),
+			createState:        defaultState,
+			createMasterSecret: defaultMasterSecret,
+			createAdminConfig:  func(require *require.Assertions, fh file.Handler) {},
+			createTfState:      defaultTfState,
+			flags:              applyFlags{},
+			wantErr:            true,
+		},
+		"no admin config file, no master secret": {
+			createConfig:       defaultConfig(cloudprovider.GCP),
+			createState:        defaultState,
+			createMasterSecret: func(require *require.Assertions, fh file.Handler) {},
+			createAdminConfig:  func(require *require.Assertions, fh file.Handler) {},
+			createTfState:      defaultTfState,
+			flags:              applyFlags{},
+			wantPhases: skipPhases{
+				skipImagePhase: struct{}{},
+				skipK8sPhase:   struct{}{},
+			},
+		},
+		"no tf state, but admin config exists": {
+			createConfig:       defaultConfig(cloudprovider.GCP),
+			createState:        defaultState,
+			createMasterSecret: defaultMasterSecret,
+			createAdminConfig:  defaultAdminConfig,
+			createTfState:      func(require *require.Assertions, fh file.Handler) {},
+			flags:              applyFlags{},
+			wantErr:            true,
+		},
+		"only config file": {
+			createConfig:       defaultConfig(cloudprovider.GCP),
+			createState:        func(require *require.Assertions, fh file.Handler) {},
+			createMasterSecret: func(require *require.Assertions, fh file.Handler) {},
+			createAdminConfig:  func(require *require.Assertions, fh file.Handler) {},
+			createTfState:      func(require *require.Assertions, fh file.Handler) {},
+			flags:              applyFlags{},
+			wantPhases: skipPhases{
+				skipImagePhase: struct{}{},
+				skipK8sPhase:   struct{}{},
+			},
+		},
+		"skip terraform": {
+			createConfig:       defaultConfig(cloudprovider.GCP),
+			createState:        defaultState,
+			createMasterSecret: func(require *require.Assertions, fh file.Handler) {},
+			createAdminConfig:  func(require *require.Assertions, fh file.Handler) {},
+			createTfState:      func(require *require.Assertions, fh file.Handler) {},
+			flags: applyFlags{
+				skipPhases: skipPhases{
+					skipInfrastructurePhase: struct{}{},
+				},
+			},
+			wantPhases: skipPhases{
+				skipInfrastructurePhase: struct{}{},
+				skipImagePhase:          struct{}{},
+				skipK8sPhase:            struct{}{},
+			},
+		},
+	}
```

Comment:
I think we should only support skip infra for self-managed cases (no tf state). Otherwise how do we ensure that a cluster upgrade is safe when a necessary infra change was skipped? Or is this always safe? At any rate this might break if a future CLI relies on a newer version infra state, but the state file migrations where not applied in prior CLI versions because infra was always skipped. Could this happen?

Reply:
Yes, skipping phases might break things further down the line, but this is something we have to accept if we want to provide the skip-phases functionality at all.
Similar to `kubeadm`, a user might have a need to skip specific phases, and we should provide the option to do so.
Simple reason why **we** would like to support it is `cdbg`: We want a way to run just the infrastructure phase, then run `cdbg`, then run anything but the infrastructure phase.